xds/internal/xdsclient/xds.go

```diff
@@ -561,6 +568,40 @@ func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger,
 	return routesRet, nil
 }
 
+func hashPoliciesProtoToSlice(policies []*v3routepb.RouteAction_HashPolicy) ([]*HashPolicy, error) {
+	var hashPoliciesRet []*HashPolicy
+	for _, p := range policies {
+		// Hash Policies are only applicable for a Ring Hash LB.
+		if !env.RingHashSupport {
+			return nil, errors.New("hash policies are only applicable for ring hash lb policy")
+		}
+		var policy HashPolicy
+		policy.Terminal = p.Terminal
+		switch p.GetPolicySpecifier().(type) {
+		case *v3routepb.RouteAction_HashPolicy_Header_:
+			policy.HashPolicyType = HashPolicyTypeHeader
+			policy.HeaderName = p.GetHeader().HeaderName
+			regex := p.GetHeader().GetRegexRewrite().GetPattern().Regex
+			re, err := regexp.Compile(regex)
+			if err != nil {
+				return nil, fmt.Errorf("hash policy %+v contains an invalid regex %q", p, regex)
+			}
+			policy.Regex = re
+			policy.RegexSubstitution = p.GetHeader().GetRegexRewrite().Substitution
+		case *v3routepb.RouteAction_HashPolicy_FilterState_:
+			if p.GetFilterState().GetKey() != "io.grpc.channel_id" {
+				return nil, fmt.Errorf("hash policy %+v contains an invalid key for filter state policy %q", p, p.GetFilterState().GetKey())
+			}
+			policy.HashPolicyType = HashPolicyTypeChannelID
+		default:
+			return nil, fmt.Errorf("hash policy %v is an unsupported hash policy", p)
```

Comment:
Print the type (with `%T`). `%v` would print only the pointer of the field, and then we don't know which policy was in the response.

Reply:
Changed to p.GetPolicySpecifier(). Crossing my fingers that that's the right thing.